### PR TITLE
Make diffing work on Windows

### PR DESCRIPTION
**Pull request by diml**
This PR changes the behavior of `(diff a b)` on Windows. Currently, it is common to compare a generated file against a source one. However, it is often the case that on Windows the generated file contains lines ending with `\r\n`, which causes the comparison to fail if the source file has unix end of lines.

After this PR, `(diff a b)` normalizes the end of lines on Windows before comparing the files. Additionally, it passes `--skip-trailing-cr` to `diff` on Windows so that `diff` does the same comparison as Dune.

In some cases, `diff` might be used to compare binary files. For this purpose a `cmp` action was added. `(cmp a b)` always compare the raw contents of `a` and `b` and never calls `diff`. Instead it simply prints:

```
Error: Files a and b differ.
```

Promoted files are still promoted as it. It's not clear what Dune should do but I expect that people working on Windows configure their git so that files are checkouted with Windows end of lines.

### CODE REVIEW

**Comment by rgrinberg** [comment]
@diml is this WIP?

**Comment by diml** [comment]
nope, I forgot to change the title

**Comment by mlasson** [comment]
Thanks, I've tested it and it seems to work fine !
(Note to other testers: you need to set-up your workspace with version dune 1.0 to notice this change of behavior).

File: src/stdune/io.ml [hunk 1]
```diff
@@ -77,10 +77,52 @@ let copy_file ~src ~dst =
...
+            len - src_pos
+        in
+        Bytes.blit_string src src_pos dst dst_pos len;
+        Bytes.sub_string dst 0 (dst_pos + len)
+      | Some i ->
+        let len = i - src_pos in
+        Bytes.blit_string src src_pos dst dst_pos len;
+        let dst_pos = dst_pos + len in
+        Bytes.set dst dst_pos '\n';
+        loop (i + 2) (dst_pos + 1)
```
Comment (rgrinberg) [inline src/stdune/io.ml]: Did you give writing this a code try with mutual recursion style? I find that much easier to read usually. I'm pretty certain that you did, since I basically learned this technique from you, but the current code is a bit hard to follow. Anyway, I've convinced myself that this is correct so approving

File: src/action.ml [hunk 2]
```diff
@@ -86,12 +88,21 @@ struct
...
-           Diff { optional = false; file1; file2 })
+           path >>= fun file2 ->
+           Syntax.get_exn Stanza.syntax >>| fun ver ->
+           let mode = if ver < (1, 0) then Diff_mode.Text_jbuild else Text in
+           Diff { optional = false; file1; file2; mode })
         ; "diff?",
           (path >>= fun file1 ->
+           path >>= fun file2 ->
+           Syntax.get_exn Stanza.syntax >>| fun ver ->
+           let mode = if ver < (1, 0) then Diff_mode.Text_jbuild else Text in
```
[the remark this answers is not in the document]
Comment (diml) [reply src/action.ml]: Agreeed. I added `Stanza.file_kind`, the advantage compared to `choose` is that it cost less when the two values require computation.

File: src/stdune/io.ml (hunk 1 above)
Comment (diml) [reply src/stdune/io.ml]: I think I did but there were a lot of variables to maintain so I thought it was better this way
